lizard_auth_server/views_api_v2.py

```diff
@@ -840,3 +844,122 @@ def form_valid(self, form):
         return HttpResponse(
             json.dumps({"user": user_data}), content_type="application/json"
         )
+
+
+class CognitoUserMigrationView(CheckCredentialsView):
+    """View to migrate users to AWS Cognito
+
+    This view is similar to CheckCredentialsView, with a few differences:
+    - users will also be found by email
+    - username and email are case-insensitive
+    - migrated users (with ``migrated_at is not None``) are ignored
+    - instead of erroring if there is a bad (or no) password, this endpoint
+      returns "password_valid": true/false.
+    - it only uses the django User model, and not the Cognito or LDAP
+      authentication backends
+    - it sets ``migrated_at`` to the current time
+    """
+
+    def form_valid(self, form):
+        """Produce data to migrate a user.
+
+        Args: See CheckCredentialsView. Password is not mandatory.
+
+        Returns:
+          A dict with keys
+          - ``user`` dict with username, email, first name, last name.
+          - ``password_valid`` bool
+
+        A 400 is "username" is missing from the request.
+
+        A 403 status if the supplied SSO_KEY/SECRET combination (Portal) does
+        not allow user migration.
+
+        A 404 status if the user does not exist
+
+        A 409 status if there are multiple users with given username/email
+        (case insensitive). A warning will be logged in this case.
+        """
+        # The JWT message is validated; now check the message's contents.
+        username = form.cleaned_data.get("username")
+        if not username:
+            return HttpResponseBadRequest("username is missing from the JWT message")
+
+        portal = Portal.objects.get(sso_key=form.cleaned_data["iss"])
+        if not portal.allow_migrate_user:
+            raise PermissionDenied("this portal is not allowed to migrate users")
+
+        # Do the authentication without the django backends, because we do not
+        # want to migrate LDAP user and we certainly do not want to do a call
+        # to Cognito, else we end up in an infinite loop.
+        try:
+            user = User.objects.get(
+                Q(username__iexact=username) | Q(email__iexact=username),
+                is_active=True,
+                user_profile__migrated_at=None,
+            )
+        except User.DoesNotExist:
+            return HttpResponseNotFound("No user found")
+        except User.MultipleObjectsReturned:
+            logger.warning("Multiple users found with username/email %s", username)
+            return HttpResponse("Multiple users found", status=409)
+
+        # Verify the password, if supplied
+        password = form.cleaned_data.get("password")
+        if password is None:
+            # Forgot password flow
+            password_valid = False  # ignored
+            migrate = True
```

Comment:
Ik dacht eerst "wat raar dat dit ook zonder wachtwoord kan", maar het is toch logisch :-)